core/node/rpc/node2nodeauth/cert.go

```diff
@@ -0,0 +1,208 @@
+package node2nodeauth
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"go.uber.org/zap"
+
+	"github.com/towns-protocol/towns/core/contracts/river"
+	. "github.com/towns-protocol/towns/core/node/base"
+	"github.com/towns-protocol/towns/core/node/crypto"
+	"github.com/towns-protocol/towns/core/node/http_client"
+	"github.com/towns-protocol/towns/core/node/nodes"
+	. "github.com/towns-protocol/towns/core/node/protocol"
+)
+
+const (
+	node2NodeCertName = "node2node"
+)
+
+var (
+	// node2NodeCertExtOID is the OID of the custom node-2-node client certificate extension.
+	node2NodeCertExtOID = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 50000, 1, 1}
+)
+
+// VerifyPeerCertificate goes through the peer certificates and verifies the node-2-node client certificate.
+// Returns nil if the certificate is valid or not found.
+func VerifyPeerCertificate(logger *zap.SugaredLogger, nodeRegistry nodes.NodeRegistry) func([][]byte, [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		for _, rawCert := range rawCerts {
+			peerCert, err := x509.ParseCertificate(rawCert)
+			if err != nil {
+				return err
+			}
+
+			if peerCert.Subject.CommonName == node2NodeCertName {
```

Comment:
Would it make sense to structure the subject common name in the certifiate to something like `<node-addr>.<operator-addr>.towns` instead of a hard coded one that is the same for all?